// @storybook/addon-console/dist/react-decorator.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

exports.reactStory = reactStory;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ReactDecorator = function (_React$Component) {
  (0, _inherits3.default)(ReactDecorator, _React$Component);

  function ReactDecorator(props) {
    (0, _classCallCheck3.default)(this, ReactDecorator);

    var _this = (0, _possibleConstructorReturn3.default)(this, (ReactDecorator.__proto__ || (0, _getPrototypeOf2.default)(ReactDecorator)).call(this, props));

    _this.props.onMount();
    return _this;
  }

  (0, _createClass3.default)(ReactDecorator, [{
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.props.onUnMount();
    }
  }, {
    key: 'render',
    value: function render() {
      return this.props.story;
    }
  }]);
  return ReactDecorator;
}(_react2.default.Component);

ReactDecorator.propTypes = {
  story: _propTypes2.default.node.isRequired,
  onMount: _propTypes2.default.func.isRequired,
  onUnMount: _propTypes2.default.func.isRequired
};
exports.default = ReactDecorator;
function reactStory(story, onMount, onUnMount) {
  return _react2.default.createElement(ReactDecorator, { story: story, onMount: onMount, onUnMount: onUnMount });
}
